### PR TITLE
io.github.shiftey.Desktop

### PR DESCRIPTION
**Pull request by advaithm**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. 
`--socket=ssh-auth` is used for cloning ssh repos
`--filesystem=host` is required for editor integration detection
`--filesystem=/opt/:ro` is also required for editor integrations 
`--talk-name=org.freedesktop.Flatpak` is required to launch editors and shells
`--filesystem=xdg-run/keyring` & `--talk-name=org.freedesktop.secrets` used to store login credentials 
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. https://github.com/shiftkey/desktop/issues/323
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned.
0001 & 0002 are purely hacks to ensure yarn is called with the --offline flag, this is not ideal for regular users
0003 changes package.json to points the dep `detect-arm64-translation` to a pre-cloned repo  stop it cloning the repo allowing us to build it completely offline. This also removed the need to patch the dep out of github-desktop. This is not required for regular users.
0004 is currently being upstreamed https://github.com/shiftkey/desktop/pull/558
[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by LeandroStanger** [comment]
bot, build com.github.Desktop

**Comment by advaithm** [comment]
bot, build com.github.Desktop

**Comment by hfiguiere** [comment]
While are `shared-modules` and `flatpak-bulder-tools` checked in? The former should just be a git submodule, and the latter like nowhere here.

**Comment by advaithm** [comment]
ah my bad. I must have messed something up while moving to this as a base. I'll readd them as submodules

**Comment by advaithm** [comment]
bot, build com.github.Desktop

File: com.github.Desktop.yaml
Comment (hfiguiere) [inline com.github.Desktop.yaml:13]:
```suggestion
  - --socket=fallback-x11
```

This assuming the app supports wayland, otherwise remove the wayland socket.

File: com.github.Desktop.metainfo.xml
Comment (hfiguiere) [inline com.github.Desktop.metainfo.xml:3]: Since this is a fork of the original upstream, it would be better to use an id associated with the fork to not misrepresent it.

File: flathub.json
Comment (hfiguiere) [inline flathub.json:3]: It's MIT licensed, so why restricting to x86_64?

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: com.github.Desktop.metainfo.xml
Comment (advaithm) [reply com.github.Desktop.metainfo.xml:3]: I will change it

File: flathub.json
Comment (advaithm) [reply flathub.json:3]: Arm is not yet supported by upstream attempting to build it for arm results in build errors. There is work being done to fix this

**Comment by hfiguiere** [comment]
Please rename the PR to reflect the Application id. It make things easier on reviewers.

**Comment by advaithm** [comment]
bot, build com.github.Desktop

File: io.github.shiftey.Desktop.metainfo.xml
Comment (hfiguiere) [inline io.github.shiftey.Desktop.metainfo.xml:3]:
```suggestion
  <id>io.github.shiftey.Desktop</id>
```

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

File: io.github.shiftey.Desktop.yaml
Comment (hfiguiere) [inline io.github.shiftey.Desktop.yaml:14]:
```suggestion
  - --socket=x11
```

I didn't say to add it if Wayland was not supported.

File: flathub.json
Comment (hfiguiere) [reply flathub.json:3]: I thought that Electron was all about being cross-platform.... *sigh*

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: io.github.shiftey.Desktop.yaml
Comment (advaithm) [reply io.github.shiftey.Desktop.yaml:14]: my bad

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

File: io.github.shiftey.Desktop.metainfo.xml
Comment (hfiguiere) [inline io.github.shiftey.Desktop.metainfo.xml:6]:
```suggestion
  <summary>GitHub Desktop is an open source GitHub app.</summary>
```
Summary doesn't need to tell what this is written in.

Also it would be better to mention this is a fork, since it is.

Comment (advaithm) [reply io.github.shiftey.Desktop.metainfo.xml]: I updated the meta info to explicitly state its a fork

**Comment by hfiguiere** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

File: flathub.json
Comment (shiftkey) [reply flathub.json:3]: > I thought that Electron was all about being cross-platform.... _sigh_

👋🏻   Upstream maintainer here: we're not supporting other architectures currently due to the native modules and Git needing to target these platforms, not Electron itself - https://github.com/shiftkey/desktop/issues/251 has more context on what's needed for ARM64 support, for example...

**Comment by nedrichards** [comment]
bot, build io.github.shiftey.Desktop

File: patches/0004-desktop-editor-intigration.patch
Comment (shiftkey) [inline patches/0004-desktop-editor-intigration.patch:16]: :art: This `console.log` can be removed from the production build

File: patches/0004-desktop-editor-intigration.patch
Comment (shiftkey) [inline patches/0004-desktop-editor-intigration.patch:36]: I have some ideas for upstream and making it easier for this `flatpak-spawn` support to not need patching like this, but first  I'd like to confirm what `process.env.FLATPAK_HOST` returns in the dev tools when you're using the test build - could you share a screenshot of that for my curiosity @advaithm ?

File: patches/0004-desktop-editor-intigration.patch
Comment (shiftkey) [inline patches/0004-desktop-editor-intigration.patch:97]: 🤔 this should use the `path.join` method from NodeJS, which will handle whether or not `path` starts with a `/`:

```ts
import { join } from 'path'

...

const pathUnderHost = join("/var/run/host", path)
```

File: patches/0004-desktop-editor-intigration.patch
Comment (shiftkey) [inline patches/0004-desktop-editor-intigration.patch:146]: 🤔 not sure if this was intended, but the `switch` statement mans these cases can all fall through to the `Alacritty` example, which matches the original code and ensures `--working-directory` passes in the user's repository:

```suggestion
+      case Shell.Gnome:
+      case Shell.Mate:
+      case Shell.Tilix:
+      case Shell.Terminator:
+      case Shell.XFCE:
+      case Shell.Alacritty:
+        return spawn("flatpak-spawn", ["--host", foundShell.path, '--working-directory', path])
```

**Review by shiftkey — COMMENTED** [review]
Some early thoughts

File: patches/0004-desktop-editor-intigration.patch
Comment (shiftkey) [inline patches/0004-desktop-editor-intigration.patch:65]: :art: this can also be removed for the production build

File: patches/0004-desktop-editor-intigration.patch
Comment (advaithm) [reply patches/0004-desktop-editor-intigration.patch:36]: Flatpak host is a env var I launch GitHub desktop with. It's value will always be 1

File: patches/0004-desktop-editor-intigration.patch
Comment (advaithm) [reply patches/0004-desktop-editor-intigration.patch:97]: Will change

File: patches/0004-desktop-editor-intigration.patch
Comment (advaithm) [reply patches/0004-desktop-editor-intigration.patch]: I'm using flatpak-spawn since it launches out side the sandbox, if you have any ideas on  how to do it without this, I'm game

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

File: io.github.shiftey.Desktop.metainfo.xml
Comment (dnohales) [inline io.github.shiftey.Desktop.metainfo.xml:24]: Let's be consistent with the indent spaces here (2 spaces).

File: io.github.shiftey.Desktop.yaml
Comment (dnohales) [inline io.github.shiftey.Desktop.yaml:79]: There are many places with 2 repeated spaces between command line arguments.

File: io.github.shiftey.Desktop.yaml
Comment (dnohales) [inline io.github.shiftey.Desktop.yaml:97]: Remove this blank line (it's not used to separate other sources)

File: patches/0004-desktop-editor-intigration.patch
Comment (dnohales) [inline patches/0004-desktop-editor-intigration.patch:209]: Maybe we could do something like this to not repeat ourselves:

```typescript
    var shellArgs = null, spawnArgs = {}
    switch (shell) {
      case Shell.Gnome:
      case Shell.Mate:
      case Shell.Tilix:
      case Shell.Terminator:
      case Shell.XFCE:
      case Shell.Alacritty:
        shellArgs = ['--working-directory', path]
        break
      case Shell.Urxvt:
        shellArgs = ['-cd', path]
        break
      case Shell.Konsole:
        shellArgs = ['--workdir', path]
      case Shell.Xterm:
        shellArgs = ['-e', '/bin/bash']
        spawnArgs = { cwd: path }
        break
      case Shell.Terminology:
        shellArgs = ['-d', path]
        break
      case Shell.Deepin:
      case Shell.Elementary:
        shellArgs = ['-w', path]
        break
      case Shell.Kitty:
        shellArgs = ['--single-instance', '--directory', path]
      default:
        return assertNever(shell, `Unknown shell: ${shell}`)
    }

    if (process.env.FLATPAK_HOST !== 'undefined') {
        return spawn("flatpak-spawn", ["--host", foundShell.path, ...shellArgs], spawnArgs)
    } else {
        return spawn(foundShell.path, shellArgs, spawnArgs)
    }
```

**Review by dnohales — COMMENTED** [review]
Nice to see this moving forward! made a little review.

File: patches/0004-desktop-editor-intigration.patch
Comment (shiftkey) [reply patches/0004-desktop-editor-intigration.patch:209]: I'm working on an upstream PR with @advaithm to make this patch easier to support in the main project (and eventually redundant) https://github.com/shiftkey/desktop/pull/555

Comment (advaithm) [reply patches/0004-desktop-editor-intigration.patch:209]: this pr has been merged, up stream should have a new release soon. I'm marking this as resolved

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Comment by nedrichards** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

File: io.github.shiftey.Desktop.metainfo.xml
Comment (barthalion) [inline io.github.shiftey.Desktop.metainfo.xml:1]: FTR this should be maintained upstream and not here.

File: io.github.shiftey.Desktop.desktop
Comment (barthalion) [inline io.github.shiftey.Desktop.desktop:1]: FTR this should be maintained upstream and not here.

File: README.md
Comment (barthalion) [inline README.md:23]: This probably could be automated to some degree, like here: https://github.com/flathub/io.openrct2.OpenRCT2/blob/master/.github/workflows/beta.yml and https://github.com/OpenRCT2/OpenRCT2/blob/develop/.github/workflows/ci.yml#L252

**Comment by barthalion** [comment]
@advaithm @shiftkey @dnohales FYI I don't see any blockers from Flathub's perspective so I'm mostly waiting for your green light to merge it.

**Comment by advaithm** [comment]
So far all issues has been fixed and anything that can be upstreamed has been. If @shiftkey is okay with the flatpak it can be merged

File: README.md
Comment (advaithm) [reply README.md:23]: I am going to look into setting a beta branch similar to what openrct 2 is doing

File: patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch
Comment (shiftkey) [inline patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: I think this patch could be simplified by upstream supporting an `OFFLINE=1` environment variable check, aside from the `findYarnVersion` being removed (defaulting to calling the `yarn` on `PATH`).

@advaithm any context on why you're not calling the local `yarn` here?

File: io.github.shiftey.Desktop.metainfo.xml
Comment (shiftkey) [inline io.github.shiftey.Desktop.metainfo.xml:22]: There's actually a 2.9.0 release that's more current than this: https://github.com/shiftkey/desktop/releases/tag/release-2.9.0-linux1

Comment (advaithm) [reply io.github.shiftey.Desktop.metainfo.xml:22]: my bad I haven't updated the metainfo file

File: patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch
Comment (advaithm) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: I believe we are, we build and install yarn in the previous step. during clean up all the files in the yarn step are removed.

File: io.github.shiftey.Desktop.metainfo.xml
Comment (shiftkey) [reply io.github.shiftey.Desktop.metainfo.xml:1]: I'd love to understand more about this, as we're using GitHub releases for some of this metadata so I'm not sure about putting that back into version control there...

File: patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch
Comment (shiftkey) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: If it's related to this version then this is ahead of what we're using in the project: `1.22.1`

https://github.com/flathub/flathub/blob/98e64ec2ff048cda986ca2d0fd9c165078f4fc20/io.github.shiftey.Desktop.yaml#L33-L36

Comment (advaithm) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: what version should I downgrade too?

Comment (shiftkey) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: The other reason why we have a local `yarn` in the project is to ensure contributors stay up to date, without necessarily needing to update their global `yarn`.

Comment (shiftkey) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: Installing a more recent version globally is fine (it needs at least a global shim to work easily on your terminal), but dropping the lookup and use of the local yarn could lead to some subtle differences in behaviour if there are updates that affect installation and usage of packages in the project.

Comment (advaithm) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: I'll look into updating the patch to call the local version of yarn

Comment (advaithm) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch:5]: I've made a patch to use the local yarn version, testing now

Comment (advaithm) [reply patches/0001-flatpak-Use-yarn-directly-with-offline-flag-in-post-.patch]: it works, moving to vendor yarn

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
@barthalion  I believe its good to merge now

**Comment by advaithm** [comment]
https://github.com/shiftkey/desktop/pull/558

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Comment by nedrichards** [comment]
bot, build io.github.shiftey.Desktop

**Comment by advaithm** [comment]
bot, build io.github.shiftey.Desktop

**Pull request closed without merging**
